Fix race condition in test_collaborating_with_completion

nathansobo opened the pull request:
The test `test_collaborating_with_completion` has a latent race condition that hasn't manifested on CI yet but could cause hangs with certain task orderings.

## The Bug

Commit `fd1494c31a` set up LSP request handlers AFTER typing the trigger character:

```rust
// Type trigger first - spawns async tasks to send completion request
editor_b.update_in(cx_b, |editor, window, cx| {
    editor.handle_input(".", window, cx);
});

// THEN set up handlers (race condition!)
fake_language_server
    .set_request_handler::<lsp::request::Completion, _, _>(...)
    .next().await.unwrap();  // Waits for handler to receive a request
```

Whether this works depends on task scheduling order, which varies by seed. If the completion request is processed before the handler is registered, the request goes to `on_unhandled_notification` which claims to handle it but sends no response, causing a hang.

## Changes

- Move handler setup BEFORE typing the trigger character
- Make `TestDispatcher::spawn_realtime` panic to prevent future non-determinism from real OS threads
- Add `execution_hash()` and `execution_count()` to TestDispatcher for debugging
- Add `DEBUG_SCHEDULER=1` logging for task execution tracing
- Document the investigation in `situation.md`

cc @localcc @SomeoneToIgnore (authors of related commits)

Release Notes:

- N/A

SomeoneToIgnore commented on crates/collab/src/tests/editor_tests.rs:
```diff
@@ -378,19 +378,10 @@ async fn test_collaborating_with_completion(cx_a: &mut TestAppContext, cx_b: &mu
-    // Return some completions from the host's language server.
-    cx_a.executor().start_waiting();
-    fake_language_server
+    // Set up the completion request handlers BEFORE typing the trigger character.
```
I think the analysis is ok, yet most of the change is nonsense.

To me, seems that we need to use `initializer` field in `FakeLspAdapter` similar to 

https://github.com/zed-industries/zed/blob/e9073eceeba8e1a71e966582566179517591ec6b/crates/collab/src/tests/editor_tests.rs#L1953-L1985

instead of all this.

SomeoneToIgnore replied:
Sorry for posting this sloppy comment, I will push see what can I apply on top and fix it further too.

nathansobo committed:
Fix race condition in test_collaborating_with_completion

The test had a race condition where LSP request handlers were registered
AFTER typing the trigger character. Whether this worked depended on task
scheduling order, which varies by seed.

Use FakeLspAdapter's initializer field to set up completion handlers
when the LSP server is created, ensuring handlers exist before any
requests can arrive.

nathansobo committed:
Simplify fix: just reorder handler setup before typing trigger

The initializer approach works but removes the explicit .next().await
synchronization that guarantees the completion request arrived and the
assertions ran. This simpler approach just reorders the existing code:
set up handlers first, then type the trigger, then await the requests.

SomeoneToIgnore committed:
Use `initializer` approach in more tests

# Conflicts:
#	crates/collab/src/tests/editor_tests.rs
